src/ucp/stream/stream_recv.c

```diff
@@ -227,32 +233,41 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
 {
     ucp_ep_ext_stream_t *ep_stream = ep->ext.stream;
     ucp_request_t       tmp_req;
+    size_t              dt_length;
     ucp_request_t       *req;
     ucs_status_ptr_t    rdesc;
     ucs_status_t        status;
+    ucp_dt_state_t      dt_state UCS_V_UNUSED; /* only for generic DT */
 
     if (UCP_DT_IS_GENERIC(datatype)) {
         ucs_error("ucp_stream_recv_nb doesn't support generic datatype");
         return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
     }
 
+    dt_length = ucp_dt_length(datatype, count, buffer, &dt_state);
+
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_unlikely(!(ep_stream->flags & UCP_EP_STREAM_FLAG_HAS_DATA))) {
-        /* No data, enqueue request and out */
+    if (ucs_likely(ucp_stream_recv_nb_is_inplace(ep_stream, dt_length))) {
+        /* We can complete request in place, use tmp request on stack */
+        req = &tmp_req;
+    } else {
         req = ucp_request_get(ep->worker);
         if (ucs_unlikely(req == NULL)) {
             req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
             goto out;
         }
-        ucp_stream_recv_request_init(req, buffer, count, datatype, cb);
-        ucs_queue_push(&ep_stream->match_q, &req->recv.queue);
-        goto ptr_out;
+
+        /* No data, enqueue request and out */
+        if (!(ep_stream->flags & UCP_EP_STREAM_FLAG_HAS_DATA)) {
+            ucp_stream_recv_request_init(req, buffer, count, dt_length,
```

Comment:
need to avoid code duplication with line 270

Reply:
This will add extra branch on fast path.